stl/inc/ranges

```diff
@@ -1915,6 +1915,104 @@ namespace ranges {
         inline constexpr _Drop_fn drop;
     } // namespace views
 
+    // CLASS TEMPLATE ranges::drop_while_view
+    // clang-format off
+    template <view _Vw, class _Pr>
+        requires input_range<_Vw> && is_object_v<_Pr> && indirect_unary_predicate<const _Pr, iterator_t<_Vw>>
+    class drop_while_view : public _Cached_position_t<forward_range<_Vw>, _Vw, drop_while_view<_Vw, _Pr>> {
+        // clang-format on
+    private:
+        /* [[no_unique_address]] */ _Vw _Range{};
+        /* [[no_unique_address]] */ _Semiregular_box<_Pr> _Pred{};
+
+    public:
+        drop_while_view() = default;
+
+        constexpr drop_while_view(_Vw _Range_, _Pr _Pred_) noexcept(
+            is_nothrow_move_constructible_v<_Vw>) // strengthened
+            : _Range(_STD move(_Range_)), _Pred{in_place, _STD move(_Pred_)} {}
+
+        _NODISCARD constexpr _Vw base() const& noexcept(
+            is_nothrow_copy_constructible_v<_Vw>) /* strengthened */ requires copy_constructible<_Vw> {
+            return _Range;
+        }
+        _NODISCARD constexpr _Vw base() && noexcept(is_nothrow_move_constructible_v<_Vw>) /* strengthened */ {
+            return _STD move(_Range);
+        }
+
+        _NODISCARD constexpr const _Pr& pred() const noexcept /* strengthened */ {
+#if _CONTAINER_DEBUG_LEVEL > 0
+            _STL_VERIFY(_Pred, "value-initialized drop_while_view has no predicate");
+#endif // _CONTAINER_DEBUG_LEVEL > 0
+            return *_Pred;
+        }
+
+        _NODISCARD constexpr auto begin() {
+#if _CONTAINER_DEBUG_LEVEL > 0
+            _STL_VERIFY(_Pred, "N4861 [range.drop.while.view] forbids calling begin on a drop_while_view that holds no "

```

Comment:
I believe this is actually correct, as Daniel said, given that this is written in code there in an implicit requirement that there is no UB so we *must* have set `_Pred` beforehand

Reply:
For `take_while_view` that is the case. The body of `take_while_view::end` is specified directly in the synopsis, which LWG considers to be the same as writing "*Effects:* Equivalent to: `{code things}`", which is "magic" thanks to http://eel.is/c++draft/structure.specifications#4. "*Effects:* Equivalent to: `{code things}`" means that the function being described inherits most of the properties defined for the equivalent code.

`pred` also doesn't need an explicit precondition for the same reason. It's specified with "*Effects:* Equivalent to: `return *pred_;`" which would have UB if `pred_` is empty (inherited from `optional::operator*`'s precondition).

For `begin`, that's not the case. `begin` has no explicit preconditions and no mechanism by which to inherit implicit preconditions. Daniel was trying to say this when he replied:

> The actual question is whether we directly or indirectly require for
both views that the corresponding member functions have a precondition
on their predicate. And the answer is: Yes, they do now with your
issue. For the Returns: element the Precondition was missing, but for
a specification it is fully defined in terms of a given inline
definition within the class it is implied by exactly that code and
there is no further specification needed, because that code says all.

but without adequate explanation.